src/main/java/oshi/hardware/platform/mac/MacSensors.java

```diff
@@ -0,0 +1,305 @@
+            return result;
+        }
+
+        service = IOKit.INSTANCE.IOServiceGetMatchingService(masterPort.getValue(),
+                IOKit.INSTANCE.IOServiceMatching("AppleSMC"));
+        if (service == 0) {
+            LOG.error("Error: no SMC found\n");
+            return result;
+        }
+
+        result = IOKit.INSTANCE.IOServiceOpen(service, SystemB.INSTANCE.mach_task_self(), 0, this.conn);
+        IOKit.INSTANCE.IOObjectRelease(service);
+        if (result != 0) {
+            LOG.error(String.format("Error: IOServiceOpen() = 0x%08x", result));
+            return result;
+        }
+        return 0;
+    }
+
+    private int smcClose() {
+        return IOKit.INSTANCE.IOServiceClose(this.conn.getValue());
+    }
+
+    /*
+     * Get Temperature from SMC
+     */
+    private double smcGetTemperature(String key, int retries) {
+        SMCVal val = new SMCVal();
+        int result = smcReadKey(key, val, retries);
+        if (result == 0) {
+            int temp = 0;
+            if (val.dataSize > 0 && Arrays.equals(val.dataType, IOKit.DATATYPE_SP78)) {
+                temp = (val.bytes[0] * 256 + val.bytes[1]) >> 2;
+            }
+            return temp / 64.0;
+        }
+        // Read failed
+        return 0d;
+    }
+
+    /*
+     * Get number of fans
+     */
+    private int smcGetFanNumber(int retries) {
+        SMCVal val = new SMCVal();
+        int result = smcReadKey(IOKit.SMC_KEY_FAN_NUM, val, retries);
+        if (result == 0) {
+            return strtoul10(val.bytes, val.dataSize);
+        }
+        // Read failed
+        return 0;
+    }
+
+    /*
+     * Get fan speed
+     */
+    private float smcGetFanSpeed(int fanNum, int retries) {
+        SMCVal val = new SMCVal();
+        String key = String.format(IOKit.SMC_KEY_FAN_SPEED, fanNum);
+        int result = smcReadKey(key, val, retries);
+        if (result == 0) {
+            return strtof(val.bytes, val.dataSize, 2);
+        }
+        // Read failed
+        return 0f;
+    }
+
+    /*
+     * Get cached keyInfo if it exists, or generate new
+     */
+    private int smcGetKeyInfo(SMCKeyData inputStructure, SMCKeyData outputStructure) {
+        if (keyInfoCache.containsKey(inputStructure.key)) {
+            outputStructure.keyInfo = keyInfoCache.get(inputStructure.key);
+        } else {
+            int result = 0;
+            inputStructure.data8 = IOKit.SMC_CMD_READ_KEYINFO;
+
+            result = smcCall(IOKit.KERNEL_INDEX_SMC, inputStructure, outputStructure);
+            if (result != 0) {
+                return result;
+            }
+        }
+        return 0;
+    }
+
+    /*
+     * Read a key from SMC
+     */
+    private int smcReadKey(String key, SMCVal val, int retries) {
+        SMCKeyData inputStructure = new SMCKeyData();
+        SMCKeyData outputStructure = new SMCKeyData();
+
+        inputStructure.key = strtoul16(key, 4);
+        int result;
+        do {
+            Util.sleep(8);
```

Comment:
Believe me, I've looked, and looked again. And slept. And drank. And looked.

You are welcome to compare... many examples out there.  I have gone line by line and tried to match function/method names.
In C:
  https://github.com/Chris911/iStats/blob/master/ext/osx_stats/smc.c (I have actually installed this ruby gem on my machine and it works flawlessly)
  https://github.com/lavoiesl/osx-cpu-temp/blob/master/smc.c
  https://github.com/beltex/libsmc/blob/master/src/smc.c
In Swift:
  https://github.com/beltex/SMCKit/blob/master/SMCKit/SMC.swift
In Objective C:
  https://github.com/FergusInLondon/SMCWrapper/blob/master/SMCWrapper/SMCWrapper.m

The only code I "skipped" is the memset() commands to zero the structure memory.  It is my understanding that JNA auto-clears structures when it creates them.  I tried manually using `.clear()` on my structures with no appreciable change.

If the buffer is overflowing it'd possibly be because the native structures and my JNA structures don't match.  I've compared them numerous times.  I'd appreciate a second set of eyes on my IOKit.java headers vs. the various smc.h structure declarations.